app/controllers/DashboardController.scala

```diff
@@ -98,20 +102,23 @@ trait DashboardController extends FrontendController with GatekeeperAuthWrapper
       val requestForm = HandleUpliftForm.form.bindFromRequest
 
       def errors(errors: Form[HandleUpliftForm]) =
-        Future.successful(Redirect(controllers.routes.DashboardController.dashboardPage))
+        fetchApplicationDetails(appId) map (details => Ok(review(errors, details)))
+
+      def recovery: PartialFunction[Throwable, play.api.mvc.Result] = {
+        case e: PreconditionFailed => {
+          Logger.warn("Rejecting the uplift failed as the application might have already been rejected.", e)
+          Redirect(controllers.routes.DashboardController.dashboardPage)
+        }
+      }
 
       def addApplicationWithValidForm(validForm: HandleUpliftForm) = {
         UpliftAction.from(validForm.action) match {
           case Some(APPROVE) =>
             applicationConnector.approveUplift(appId, loggedIn.get) map (
-              ApproveUpliftSuccessful => Redirect(controllers.routes.DashboardController.dashboardPage)) recover {
-              case e:ApproveUpliftPreconditionFailed => {
-                Logger.warn("Request to uplift application failed as application might have already been uplifted.", e)
-                Redirect(controllers.routes.DashboardController.dashboardPage)
-              }
-            }
+              ApproveUpliftSuccessful => Redirect(controllers.routes.DashboardController.dashboardPage)) recover recovery
           case Some(REJECT) =>
-            Future.successful(Redirect(controllers.routes.DashboardController.dashboardPage))
+            applicationConnector.rejectUplift(appId, loggedIn.get, validForm.reason.get) map (
+              RejectUpliftSuccessful => Redirect(controllers.routes.DashboardController.dashboardPage)) recover recovery
         }
       }
```

Comment:
What happens here if `UpliftAction.from` returns a None?  I assume that results in the 'Technical Difficulties' 500 page being shown.